engine/execution/computation/computer/uploader/uploader_test.go

```diff
@@ -95,35 +99,75 @@ func Test_AsyncUploader(t *testing.T) {
 		require.Equal(t, 3, callCount)
 	})
 
-	time.Sleep(1 * time.Second)
-
+	// This test shuts down the async uploader right after the upload has started. The upload has an error to force
+	// the retry mechanism to kick in (under normal circumstances). Since the component is shutting down, the retry
+	// should not kick in.
+	//
+	// sequence of events:
+	// 1. create async uploader and initiate upload with an error - to force retrying
+	// 2. shut down async uploader right after upload initiated (not completed)
+	// 3. assert that upload called only once even when trying to use retry mechanism
 	t.Run("stopping component stops retrying", func(t *testing.T) {
 		testutils.SkipUnless(t, testutils.TEST_FLAKY, "flaky")
 
 		callCount := 0
+		t.Log("test started grID:", string(bytes.Fields(debug.Stack())[1]))
 
-		wg := sync.WaitGroup{}
-		wg.Add(1)
+		// this wait group ensures that async uploader has a chance to start the upload before component is shut down
+		// otherwise, there's a race condition that can happen where the component can shut down before the async uploader
+		// has a chance to start the upload
+		wgUploadStarted := sync.WaitGroup{}
+		wgUploadStarted.Add(1)
+
+		// this wait group ensures that async uploader won't send an error (to test if retry will kick in) until
+		// the component has initiated shutting down (which should stop retry from working)
+		wgShutdownStarted := sync.WaitGroup{}
+		wgShutdownStarted.Add(1)
+		t.Log("added 1 to wait group grID:", string(bytes.Fields(debug.Stack())[1]))
 
 		uploader := &DummyUploader{
 			f: func() error {
-				defer func() {
+				t.Log("DummyUploader func() - about to call wgUploadStarted.Done() grID:", string(bytes.Fields(debug.Stack())[1]))
+				// signal to main goroutine that upload started, so it can initiate shutting down component
+				wgUploadStarted.Done()
+
+				t.Log("DummyUpload func() waiting for component shutdown to start grID:", string(bytes.Fields(debug.Stack())[1]))
+				wgShutdownStarted.Wait()
+				t.Log("DummyUploader func() component shutdown started, about to return error grID:", string(bytes.Fields(debug.Stack())[1]))
+
+				// force an upload error to test that upload is never retried (because component is shut down)
+				// normally, we would see retry mechanism kick in and the callCount would be > 1
+				// but since component has started shutting down, we expect callCount to be 1
+				if callCount < 5 {
```

Comment:
Why is this less than 5? Shouldn't this be called only once really?

Reply:
Yeah, the comment also doesn't explain why it's 5, from reading the comment, i would assume it'd be `callCount <= 1`

Reply:
yes, this SHOULD be called only once - but we want the test to TRY and call it more than once to prove that it was only called once. If we changed it to `callCount < 1` that wouldn't prove that the test tried to call it more than once and wouldn't prove that stopping the component stopped the retry mechanism.

Reply:
added clarification comments to explain this
https://github.com/onflow/flow-go/pull/2453/commits/a6271a474bd1293fe7417f32b40ad2377de12b36